doggos/knowledge/rule.py

```diff
@@ -1,24 +1,105 @@
+from doggos.knowledge.consequents.takagi_sugeno_consequent import TakagiSugenoConsequent
+from doggos.knowledge.consequents.mamdani_consequent import MamdaniConsequent
 
 
 class Rule:
+    """
+    Class used to represent a fuzzy rule,
+    that is used in fuzzy logic systems to infer an output based on input variables.
+    It consists from an antecedent(premise) and a consequent.
+    https://en.wikipedia.org/wiki/Fuzzy_rule
+
+    Attributes
+    --------------------------------------------
+
+    __antecedent : Antecedent
+        Antecedent of the rule. Has getter and setter.
+
+    __consequent : Consequent
+        Consequent of the rule. Has getter and setter.
+
+    Methods
+    --------------------------------------------
+    output(self, features: Dict[str, float], measures: Dict[str, float] = None) -> OutputType:
+        Calculate firing value and output of rule, then return output of type that is depending on type of consequent.
+
+    Examples:
+    --------------------------------------------
+
+    """
+
     def __init__(self, antecedent: Antecedent, consequent: Consequent):
-        self.__firing_value: MembershipDegree = None
-        self.__rule_output: List[MembershipDegree] or float = None
-        self.__antecedent = antecedent
-        self.__consequent = consequent
+        """
+        Create a new rule with defined antecedent and consequent.
+
+        :param antecedent: antecedent of the rule
+        :param consequent: consequent of the rule
+        """
+        self.__antecedent = None
+        self.__consequent = None
+        if isinstance(antecedent, Antecedent):
+            self.__antecedent = antecedent
+        if isinstance(consequent, Consequent):
+            self.__consequent = consequent
 
     @property
     def antecedent(self) -> Antecedent:
-        pass
+        return self.__antecedent
+
+    @antecedent.setter
+    def antecedent(self, new_antecedent: Antecedent) -> NoReturn:
+        """Sets firing value and rule output to None"""
+        if isinstance(new_antecedent, Antecedent):
+            self.__antecedent = new_antecedent
+        else:
+            raise ValueError("antecedent of rule must be set to instance of Antecedent class")
 
     @property
     def consequent(self) -> Consequent:
-        pass
+        return self.__consequent
+
+    @consequent.setter
+    def consequent(self, new_consequent: Consequent) -> NoReturn:
+        """Sets rule output to None"""
+        if isinstance(new_consequent, Consequent):
+            self.__consequent = new_consequent
+        else:
+            raise ValueError("consequent of rule must be set to instance of Consequent class")
+
+    """Types returned by Output"""
+    MamdaniOutputType = NewType('MamdaniOutputType', List[MembershipDegree])
+    TakagiSugenoOutputType = NewType('TakagiSugenoOutputType', float)
+    OutputType = NewType('OutputType', MamdaniOutputType or TakagiSugenoOutputType)
```

Comment:
It needs to be discussed, looks like nice idea, but needs to be adjusted a bit since for TakagiSugeno you return not only float, but also firing.